website/addons/box/utils.py

```diff
@@ -0,0 +1,199 @@
+                        'box_download', path=cleaned_path)
+                },
+                'path': cleaned_path,
+            })
+        if extra:
+            params.update(extra)
+        # Prefix the action with box_
+        self.node.add_log(
+            action="box_{0}".format(action),
+            params=params,
+            auth=self.auth
+        )
+        if save:
+            self.node.save()
+
+
+def is_subdir(path, directory):
+    if not (path and directory):
+        return False
+    #make both absolute
+    abs_directory = os.path.abspath(directory).lower()
+    abs_path = os.path.abspath(path).lower()
+    return os.path.commonprefix([abs_path, abs_directory]) == abs_directory
+
+
+def is_authorizer(auth, node_addon):
+    """Return if the auth object's user is the same as the authorizer of the node."""
+    return auth.user == node_addon.user_settings.owner
+
+
+def abort_if_not_subdir(path, directory):
+    """Check if path is a subdirectory of directory. If not, abort the current
+    request with a 403 error.
+    """
+    if not is_subdir(clean_path(path), clean_path(directory)):
+        raise HTTPError(http.FORBIDDEN)
+    return True
+
+
+def get_file_name(path):
+    """Given a path, get just the base filename.
+    Handles "/foo/bar/baz.txt/" -> "baz.txt"
+    """
+    return os.path.basename(path.strip('/'))
+
+
+def clean_path(path):
+    """Ensure a path is formatted correctly for url_for."""
+    if path is None:
+        return ''
+    return path.strip('/')
+
+
+def make_file_response(fileobject, metadata):
+    """Builds a response from a file-like object and metadata returned by
+    a Box client.
+    """
+    resp = make_response(fileobject.read())
+    filename = get_file_name(metadata['path'])
+    rev = metadata.get('rev')
+    if rev:
+        # add revision to filename
+        # foo.mp3 -> foo-abc123.mp3
+        filename = '-{rev}'.format(rev=rev).join(os.path.splitext(filename))
+    disposition = 'attachment; filename={0}'.format(filename)
+    resp.headers['Content-Disposition'] = disposition
+    resp.headers['Content-Type'] = metadata.get('mime_type', 'application/octet-stream')
+    return resp
+
+
+def ensure_leading_slash(path):
+    if not path.startswith('/'):
+        return '/' + path
+    return path
+
+
+def build_box_urls(item, node):
+    path = clean_path(item['path'])  # Strip trailing and leading slashes
+    if item['type']==u'folder':
+        return {
+            'upload': node.api_url_for('box_upload', path=path),
+            # Endpoint for fetching all of a folder's contents
+            'fetch': node.api_url_for('box_hgrid_data_contents', path=path),
+            # Add extra endpoint for fetching folders only (used by node settings page)
+            # NOTE: querystring params in camel-case
+            'folders': node.api_url_for('box_hgrid_data_contents',
+                path=path, foldersOnly=1)
+        }
+    else:
+        return {
```

Comment:
This should never be run, remove it and for safety add at the top of the function:
```python
assert item['type'] == 'folder', 'Can only build urls for box folders'
```